File: src/infrastructure/csvparser/CSVParser.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using core.fs;
using core.fs.Adapters.CSV;
using CsvHelper;
using CsvHelper.Configuration;
using Microsoft.FSharp.Core;

namespace csvparser
{
    public class CSVParser : ICSVParser
    {
        public FSharpResult<IEnumerable<T>,ServiceError> Parse<T>(string content)
        {
            try
            {
                using var reader = new StringReader(content);
                var config = new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    PrepareHeaderForMatch = (header) =>
                    {
                        // remove spaces and capitalize the first letter so that
                        // CSVs produced by various services like Coinbase, TD Ameritrade, etc.
                        // can be parsed correctly
                        var spacesRemoved = Regex.Replace(header.Header, @"\s", string.Empty);
                        return Char.ToUpper(spacesRemoved[0]) + spacesRemoved.ToLower().Substring(1);
                    },
                    ShouldSkipRecord = (arr) => arr.Row.ColumnCount == 0 || arr.Row.GetField(0) == "***END OF FILE***"
                    // TD Ameritrade includes **END OF FILE** at the end
                };
                    
                var csvReader = new CsvReader(reader, config);

                var records = csvReader.GetRecords<T>().ToList(); // doing to list to make sure that the file is read
                // before reader gets disposed
                return FSharpResult<IEnumerable<T>, ServiceError>.NewOk(records);
            }
            catch(HeaderValidationException ex)
            {
                var error = "Header validation failed: " + ex.Message;
                return FSharpResult<IEnumerable<T>, ServiceError>.NewError(new ServiceError(error));
            }
        }
    }
}

File: src/infrastructure/csvparser.fs/CSVParser.fs
namespace csvparser.fs

open System
open System.IO
open System.Globalization
open System.Text.RegularExpressions
open CsvHelper
open CsvHelper.Configuration
open core.fs
open core.fs.Adapters.CSV

type CSVParser() =
    interface ICSVParser with
        member _.Parse<'T>(content: string) : Result<seq<'T>, ServiceError> =
            try
                use reader = new StringReader(content)
                let config = 
                    CsvConfiguration(
                        CultureInfo.InvariantCulture,
                        PrepareHeaderForMatch = (fun args ->
                            // remove spaces and capitalize the first letter so that
                            // CSVs produced by various services like Coinbase, TD Ameritrade, etc.
                            // can be parsed correctly
                            let spacesRemoved = Regex.Replace(args.Header, @"\s", String.Empty)
                            if spacesRemoved.Length > 0 then
                                Char.ToUpper(spacesRemoved.[0]).ToString() + spacesRemoved.ToLower().Substring(1)
                            else
                                spacesRemoved),
                        ShouldSkipRecord = (fun args ->
                            args.Row.ColumnCount = 0 || args.Row.GetField(0) = "***END OF FILE***")
                            // TD Ameritrade includes ***END OF FILE*** at the end
                    )
                
                use csvReader = new CsvReader(reader, config)
                
                // doing to list to make sure that the file is read
                // before reader gets disposed
                let records = csvReader.GetRecords<'T>() |> Seq.toList
                Ok (records :> seq<'T>)
            with
            | :? HeaderValidationException as ex ->
                let error = "Header validation failed: " + ex.Message
                Error (ServiceError(error))
            | ex ->
                Error (ServiceError(ex.Message))
